### PR TITLE
Adding retries on HTTP errors

### PR DESCRIPTION
**Pull request by vladbagmet**
### Rationale
Currently API client does not make retries on server errors. This PR changes that.

### Changes
*Describe what you changed*

[JIRA ticket](put the link to the corresponding JIRA ticket here)

#### TODO
- [ ] Update CHANGELOG.md

### CODE REVIEW

**Review by katsanva — CHANGES_REQUESTED** [review]
We have removed this functionality from node SDK due to API bombing, was this change aligned with all of the teams?
@ville can you please give your input?

**Comment by ville** [comment]
I think this should not be on by default.

**Comment by vladbagmet** [comment]
> I think this should not be on by default.

@ville, I disabled it by default. Now users can set their own limit once api client is initialised.

File: src/corva/api.py [hunk 1]
```diff
@@ -14,6 +16,14 @@ class Api:
     """
 
     TIMEOUT_LIMITS = (3, 30)  # seconds
+    DEFAULT_MAX_RETRIES = 0
```
Comment (oleksii-symon-corva-ai) [inline src/corva/api.py:19]:
```suggestion
    DEFAULT_MAX_RETRIES = int(0)
```

File: src/corva/api.py [hunk 2]
```diff
@@ -31,28 +41,39 @@ def __init__(
...
     def default_headers(self):
         return {
-            'Authorization': f'API {self.api_key}',
-            'X-Corva-App': self.app_key,
+            "Authorization": f"API {self.api_key}",
+            "X-Corva-App": self.app_key,
         }
 
+    @property
+    def max_retries(self):
```
Comment (oleksii-symon-corva-ai) [inline src/corva/api.py:54]:
```suggestion
    def max_retries(self) -> int:
```

File: src/corva/api.py [hunk 3]
```diff
@@ -31,28 +41,39 @@ def __init__(
...
+        return self._max_retries
+
+    @max_retries.setter
+    def max_retries(self, value):
+        if not (isinstance(value, int) and 0 <= value <= 10):
```
Comment (oleksii-symon-corva-ai) [inline src/corva/api.py:59]:
```suggestion
    def max_retries(self, value: int) -> None:
        if not (0 <= value <= 10):
```

File: src/corva/api.py [hunk 4]
```diff
@@ -113,22 +135,27 @@ def _request(
...
 
-        response = requests.request(
-            method=method,
-            url=url,
-            params=params,
-            json=data,
-            headers=headers,
-            timeout=timeout,
-        )
+        for retry_attempt in range(max(self._max_retries, 1)):
```
Comment (oleksii-symon-corva-ai) [inline src/corva/api.py:138]: If _max_retries == 1 - at most single request will be made

**Comment by vladbagmet** [comment]
> Did you think about using popular libraries for retries like `tenacity` or `stamina`? We have single place, where we use retries - but still maybe it would lead to less logic for maintenance
> 
> Also, do we want to add this feature to docs for everyone to see?

@oleksii-symon-corva-ai, I've updated the code to make tenacity handle retries. Also, updated docs. Please review this PR again.

File: docs/modules/ROOT/pages/index.adoc [hunk 5]
```diff
@@ -269,6 +269,22 @@ include::example$api/tutorial007.py[]
...
+* 428
+* 500
+* 502
+* 503
+* 504
+
+[source,python]
+----
+include::example$api/tutorial008.py[]
+
```
Comment (oleksii-symon-corva-ai) [inline docs/modules/ROOT/pages/index.adoc:287]:
```suggestion
----
include::example$api/tutorial008.py[]
---
```

Comment (oleksii-symon-corva-ai) [reply docs/modules/ROOT/pages/index.adoc:287]: You can use `make docs-view` to verify docs correctness

Remember to use `branches: HEAD` here https://github.com/corva-ai/python-sdk/blob/master/docs/antora-playbook.yml#L9 before running that command

File: src/corva/api.py [hunk 6]
```diff
@@ -113,22 +171,46 @@ def _request(
...
-        )
+        if self.max_retries > 0:
+            retry_decorator = retry(
+                stop=stop_after_attempt(self.max_retries),
+                wait=wait_random_exponential(multiplier=0.25, max=10),
+                retry=retry_if_result(
+                    lambda r: r.status_code in retryable_status_codes
+                ),
+            )
+            retrying_request = retry_decorator(self._execute_request)
```
Comment (oleksii-symon-corva-ai) [inline src/corva/api.py:182]: Would retrying code block here be simpler? https://tenacity.readthedocs.io/en/latest/#retrying-code-block

e.g. there would be no need to add separate `_execute_request` method

File: tests/unit/test_docs/test_api.py [hunk 7]
```diff
@@ -120,3 +121,12 @@ def test_tutorial007(app_runner, mocker: MockerFixture):
...
     time_mock.assert_called_once()
+
+
+def test_tutorial008(app_runner, mocker: MockerFixture):
+    time_event = ScheduledDataTimeEvent(
+        asset_id=0, company_id=0, start_time=0, end_time=0
+    )
+    time_mock = mocker.patch.object(Api, 'insert_data')
+    app_runner(tutorial008.scheduled_app, time_event)
+    time_mock.assert_not_called()
```
Comment (oleksii-symon-corva-ai) [inline tests/unit/test_docs/test_api.py:132]:
```suggestion
    app_runner(tutorial008.scheduled_app, time_event)
```

**Review by oleksii-symon-corva-ai — COMMENTED** [review]
Great job, only small comments

**Comment by oleksii-symon-corva-ai** [comment]
Also, remember to update `Unreleased` section `CHANGELOG.md` with your feature

File: tests/unit/test_docs/test_api.py (hunk 7 above)
Comment (vladbagmet) [reply tests/unit/test_docs/test_api.py]: This [is addressed](https://github.com/corva-ai/python-sdk/pull/81/commits/fe5a3dfda5454d40892c4e5d386114ec1db2943c).

File: src/corva/api.py (hunk 6 above)
Comment (vladbagmet) [reply src/corva/api.py:182]: To be honest it was simpler before using 3-rd party retrying lib and the code we had initially was very similar to https://tenacity.readthedocs.io/en/latest/#retrying-code-block. I haven't managed make that code in example work without breaking anything.

**Comment by vladbagmet** [comment]
> Also, remember to update `Unreleased` section in `CHANGELOG.md` with your feature

I've updated `Unreleased` section in `CHANGELOG.md` within [#fe5a3df](https://github.com/corva-ai/python-sdk/pull/81/commits/fe5a3dfda5454d40892c4e5d386114ec1db2943c).

File: CHANGELOG.md [hunk 8]
```diff
@@ -5,6 +5,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+-  Added functionality to make retries for the failed HTTP requests.
```
Comment (oleksii-symon-corva-ai) [inline CHANGELOG.md:8]:
```suggestion
## [Unreleased]

### Added
-  Functionality to make retries for the failed HTTP requests.
```

**Review by oleksii-symon-corva-ai — APPROVED** [review]

**Comment by vladbagmet** [comment]
@oleksii-symon-corva-ai, thank you for the review!

**Pull request merged**